staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // Run spawns the secure http server. It only returns if stopCh is closed
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
-	delayedStopCh := make(chan struct{})
+	events := s.TerminationEvents
+
+	signalFn := s.overrideTerminationEventSignalFunc
+	if signalFn == nil {
+		signalFn = func() {
+			// wait until we receive a KILL signal
+			<-stopCh
+			events.WithName(ShutdownInitiated).Signal()
+
+			// fire AfterShutdownDelayDuration event when ShutdownDelayDuration elapses
+			<-time.After(s.ShutdownDelayDuration)
+			events.WithName(AfterShutdownDelayDuration).Signal()
+
+			// wait until all existing request(s) in flight that we account for are drained.
+			s.HandlerChainWaitGroup.Wait()
+			events.WithName(InFlightRequestsDrained).Signal()
+		}
+	}
+	go signalFn()
 
 	go func() {
-		defer close(delayedStopCh)
-
-		<-stopCh
-
 		// As soon as shutdown is initiated, /readyz should start returning failure.
-		// This gives the load balancer a window defined by ShutdownDelayDuration to detect that /readyz is red
-		// and stop sending traffic to this server.
-		close(s.readinessStopCh)
+		// This gives the load balancer a window defined by ShutdownDelayDuration to
+		// detect that /readyz is red and stop sending traffic to this server.
+		defer close(s.readinessStopCh)
 
-		time.Sleep(s.ShutdownDelayDuration)
+		<-events.WithName(ShutdownInitiated).Signaled()
 	}()
 
-	// close socket after delayed stopCh
-	stoppedCh, err := s.NonBlockingRun(delayedStopCh)
+	// Start the audit backend before any request comes in. This means we must call Backend.Run
+	// before http server start serving. Otherwise the Backend.ProcessEvents call might block.
+	if s.AuditBackend != nil {
+		if err := s.AuditBackend.Run(events.WithName(InFlightRequestsDrained).Signaled()); err != nil {
+			return fmt.Errorf("failed to run the audit backend: %v", err)
+		}
+	}
+
+	serverShutdownCh, listenerStoppedCh, err := s.NonBlockingRun(events.WithName(AfterShutdownDelayDuration).Signaled())
 	if err != nil {
 		return err
 	}
-
-	drainedCh := make(chan struct{})
 	go func() {
-		defer close(drainedCh)
-
-		// wait for the delayed stopCh before closing the handler chain (it rejects everything after Wait has been called).
-		<-delayedStopCh
-
-		// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
-		s.HandlerChainWaitGroup.Wait()
+		<-listenerStoppedCh
+		events.WithName(HTTPServerStoppedListening).Signal()
 	}()
 
+	klog.V(1).Info("[graceful-termination] waiting for shutdown to be initiated")
 	<-stopCh
 
 	// run shutdown hooks directly. This includes deregistering from the kubernetes endpoint in case of kube-apiserver.
 	err = s.RunPreShutdownHooks()
 	if err != nil {
 		return err
 	}
+	klog.V(1).Info("[graceful-termination] RunPreShutdownHooks has completed")
```

Comment:
should we convert to structured logging?

Reply:
`[graceful-termination]` is just a prefix so we can search all graceful termination related log entries with this keyword. Is there a way to add a prefix in structured logging?